src/library/coverart.cpp

```diff
@@ -28,20 +31,69 @@ QString typeToString(CoverInfo::Type type) {
     return "INVALID TYPE VALUE";
 }
 
+QString coverInfoRelativeToString(const CoverInfoRelative& infoRelative) {
+    return typeToString(infoRelative.type) % QLatin1Literal(",") %
+           sourceToString(infoRelative.source) % QLatin1Literal(",") %
+           infoRelative.coverLocation % QLatin1Literal(",") %
+           QLatin1Literal("0x") % QString::number(infoRelative.hash, 16);
+}
+
+QString coverInfoToString(const CoverInfo& info) {
+    return coverInfoRelativeToString(info) % QLatin1Literal(",") %
+           info.trackLocation % QLatin1Literal(",");
+}
+} // anonymous namespace
+
+
+bool operator==(const CoverInfoRelative& a, const CoverInfoRelative& b) {
+    return a.source == b.source &&
+            a.type == b.type &&
+            a.coverLocation == b.coverLocation &&
+            a.hash == b.hash;
+}
+
+bool operator!=(const CoverInfoRelative& a, const CoverInfoRelative& b) {
+    return !(a == b);
+}
+
+QDebug operator<<(QDebug dbg, const CoverInfoRelative& infoRelative) {
+    return dbg.maybeSpace() << QString("CoverInfoRelative(%1)")
+            .arg(coverInfoRelativeToString(infoRelative));
+}
+
+bool operator==(const CoverInfo& a, const CoverInfo& b) {
+    return static_cast<const CoverInfoRelative&>(a) ==
+                    static_cast<const CoverInfoRelative&>(b) &&
+            a.trackLocation == b.trackLocation;
+}
+
+bool operator!=(const CoverInfo& a, const CoverInfo& b) {
+    return !(a == b);
+}
+
 QDebug operator<<(QDebug dbg, const CoverInfo& info) {
-    return dbg.maybeSpace() << QString("CoverInfo(%1,%2,%3,%4,%5)")
-            .arg(typeToString(info.type),
-                 sourceToString(info.source),
-                 info.coverLocation,
-                 QString::number(info.hash), 
-                 info.trackLocation);
+    return dbg.maybeSpace() << QString("CoverInfo(%1)")
+            .arg(coverInfoToString(info));
+}
+
+bool operator==(const CoverArt& a, const CoverArt& b) {
+    // Only count image in the equality if both are non-null.
```

Comment:
Since the operator== is not used, it is hard to predict how it will be used. 

Because the hash is not significant enough here, and it depends on the usage of == if this is true after the fix, I will just remove this operator for now.